src/web/debugger/session.ts

```diff
@@ -0,0 +1,296 @@
+
+        this.sendResponse(response);
+    }
+
+    protected continueRequest(response: DebugProtocol.ContinueResponse, args: DebugProtocol.ContinueArguments): void {
+        this.driver.resume(SimulatorDebuggerCommand.Resume);
+        this.sendResponse(response);
+    }
+
+    protected nextRequest(response: DebugProtocol.NextResponse, args: DebugProtocol.NextArguments): void {
+        this.driver.resume(SimulatorDebuggerCommand.StepOver);
+        this.sendResponse(response);
+    }
+
+    protected stepInRequest(response: DebugProtocol.StepInResponse, args: DebugProtocol.StepInArguments): void {
+        this.driver.resume(SimulatorDebuggerCommand.StepInto);
+        this.sendResponse(response);
+    }
+
+    protected stepOutRequest(response: DebugProtocol.StepOutResponse, args: DebugProtocol.StepOutArguments): void {
+        this.driver.resume(SimulatorDebuggerCommand.StepOut);
+        this.sendResponse(response);
+    }
+
+    protected pauseRequest(response: DebugProtocol.PauseResponse, args: DebugProtocol.PauseArguments): void {
+        this.driver.resume(SimulatorDebuggerCommand.Pause);
+        this.sendResponse(response);
+    }
+
+    protected threadsRequest(response: DebugProtocol.ThreadsResponse): void {
+        response.body = { threads: [{ id: SimDebugSession.THREAD_ID, name: "main" }] }
+        this.sendResponse(response);
+    }
+
+    protected stackTraceRequest(response: DebugProtocol.StackTraceResponse, args: DebugProtocol.StackTraceArguments): void {
+        if (this.lastBreak && this.state) {
+            const frames = this.state.getFrames();
+            response.body = { stackFrames: frames };
+        }
+        this.sendResponse(response);
+    }
+
+    protected scopesRequest(response: DebugProtocol.ScopesResponse, args: DebugProtocol.ScopesArguments): void {
+        if (this.state) {
+            response.body = { scopes: this.state.getScopes(args.frameId) }
+        }
+
+        this.sendResponse(response);
+    }
+
+    protected async variablesRequest(response: DebugProtocol.VariablesResponse, args: DebugProtocol.VariablesArguments) {
+        if (this.state) {
+            const variables = await this.state.getVariables(args.variablesReference);
+            response.body = { variables };
+        }
+
+        this.sendResponse(response);
+    }
+
+    private onDebuggerBreakpoint(breakMsg: pxsim.DebuggerBreakpointMessage) {
+        this.lastBreak = breakMsg;
+        this.state = new StoppedState(this.lastBreak, this.breakpoints!, this.driver);
+
+        if (breakMsg.exceptionMessage) {
+            const message = breakMsg.exceptionMessage.replace(/___\d+/g, '');
+            this.sendEvent(new StoppedEvent("exception", SimDebugSession.THREAD_ID, message));
+        }
+        else {
+            this.sendEvent(new StoppedEvent("breakpoint", SimDebugSession.THREAD_ID));
+        }
+    }
+
+    private onDebuggerWarning(warnMsg: pxsim.DebuggerWarningMessage) {
```

Comment:
Was it intentional to leave this one empty?

Reply:
yes... i added it because we surface it in simdriver in pxt but I don't actually know what these messages are or if they are generated anywhere. couldn't find in pxt where it gets posted (though maybe some targets use it). i'll add a comment